Support bitfield allocation units larger than 64 bits

glyn opened the pull request:
Individual bitfields are still limited to at most 64 bits, but this restriction can be weakened when Rust supports `u128`.

This implements issue #816.

Usage notes:

* Since common code is added to each generated binding, a program which uses
  more than one binding may need to work around the duplication by including
  each binding in its own module.
* The values created by bitfield allocation unit constructors can be assigned
  directly to the corresponding struct fields with no need for transmutation.

Implementation notes:

`__BindgenBitfieldUnit` represents a bitfield allocation unit using a `Storage`
type accessible as a slice of `u8`. The alignment of the unit is inherited from
an `Align` type by virtue of the field:
```
align: [Align; 0],
```
The position of this field in the struct is irrelevant.

It is assumed that the alignment of the `Storage` type is no larger than the
alignment of the `Align` type, which will be true if the `Storage` type is, for
example, an array of `u8`. This assumption is checked in a debug assertion.

Although the double underscore (__) prefix is reserved for implementations of
C++, there are precedents for this convention elsewhere in bindgen and so the
convention is adopted here too.

Acknowledgement:

Thanks to @fitzgen for an initial implementation of `__BindgenBitfieldUnit` and
code to integrate it into bindgen.


r? @emilio

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

emilio commented on src/codegen/bitfield_unit.rs:
```diff
@@ -0,0 +1,83 @@
+#[derive(Copy, Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct BindgenBitfieldUnit<Storage, Align>
```
This definitely needs to be `repr(C)`.

emilio replied:
Also, I'd recommend maybe to put a double underscore as the prefix for the name, since that's _technically_ reserved in C. Not sure what's @fitzgen's opinion on that.

glyn replied:
@emilio Nice catch! Thanks. Fixed.

glyn commented:
@emilio Thanks for the feedback. I added `repr(C)` and the double underscore prefix. Since @fitgen doesn't seem to be around at the moment and you have begun a review, I've changed the reviewer to you. Let's hope @highfive notices...

glyn commented:
r? @emilio

emilio commented on src/codegen/bitfield_unit.rs:
```diff
@@ -0,0 +1,84 @@
+{
+    #[inline]
+    pub fn new(storage: Storage) -> Self {
+        use std::mem::align_of;
```
This somehow needs to be `core` when `--use-core` is passed.

glyn replied:
@emilio The simplest solution is to remove the assertion (and its use of `align_of`) as it's not needed currently and it's very unlikely that anyone will trip over the assumption in the future.

emilio commented on src/codegen/bitfield_unit.rs:
```diff
@@ -0,0 +1,84 @@
+    #[inline]
+    pub fn new(storage: Storage) -> Self {
+        use std::mem::align_of;
+        debug_assert!(align_of::<Storage>() <= align_of::<Align>());
```
This can probably be a release `assert!`. It will go away anyway.

glyn replied:
@emilio The simplest solution is to remove the assertion as it's not needed currently and it's very unlikely that anyone will trip over the assumption in the future.

emilio commented on src/features.rs:
```diff
@@ -140,8 +140,6 @@ macro_rules! rust_feature_def {
     /// Untagged unions ([RFC 1444](https://github.com/rust-lang/rfcs/blob/master/text/1444-union.md))
     => untagged_union;
-    /// Constant function ([RFC 911](https://github.com/rust-lang/rfcs/blob/master/text/0911-const-fn.md))
-    => const_fn;
```
I think `const fn` was also needed for SpiderMonkey, so not sure we can get away without it so easily...

glyn replied:
@emilio I'll leave it out for now because it would be easy to add back in. Certainly the bitfield allocation unit constructors have no hope of being `const fn`.

emilio commented on bindgen-integration/cpp/Test.cc:
```diff
@@ -83,22 +118,80 @@ Fifth::assert(unsigned short nWeekDay,
-      this->nMonthDay == nMonthDay &&
-      this->nMonth == nMonth &&
-      this->byte == byte;
+    printf("\n");
```
Not sure how worth are the printfs? i'd usually just use `gdb` for this.

glyn replied:
@emilio Agreed. Better to delete this clutter. It's trivial to add back in for debugging if someone doesn't have `gdb`.

emilio commented:
I took a closer look, and I can't think of why this shouldn't get merged. Thanks a lot @glyn!

@bors-servo r+

bors-servo commented:
:pushpin: Commit f0e0531 has been approved by `emilio`

<!-- @bors-servo r=emilio f0e05310b43e88e541ed011d20994c02fdcc1a3a -->

bors-servo commented:
:hourglass: Testing commit f0e05310b43e88e541ed011d20994c02fdcc1a3a with merge 7c3584d7bc328a01ae5352e191f1fd41bb7fb1a9...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang-nursery/rust-bindgen/builds/306249229?utm_source=github_status&utm_medium=notification)
Approved by: emilio
Pushing 7c3584d7bc328a01ae5352e191f1fd41bb7fb1a9 to master...

fitzgen commented:
Thanks for taking this over the finishing line @glyn!

We should probably go back and re-test all of our bitfield bugs and see if we fixed any of them with this change.
